File: examples/php/crud_lifecycle.php
<?php
/**
 * End-to-end CRUD lifecycle test for the Pescheck PHP client.
 *
 * Exercises the full v2 surface against the API and self-cleans where possible:
 *   checks      list -> retrieve one
 *   profiles    create -> retrieve -> patch -> delete
 *   screenings  create (needs a profile + candidate) -> retrieve -> documents
 *   webhooks    create -> list -> delete
 *   divisions   list -> reuse-or-create by name -> patch
 *
 * Prints a line per step and "E2E CRUD OK" on success; exits non-zero on error.
 *
 * Env:
 *   PESCHECK_BASE_URL      e.g. https://api-staging.pescheck.io
 *   PESCHECK_ACCESS_TOKEN  OAuth2 client-credentials access token
 *   PESCHECK_TEST_EMAIL    candidate email for the screening (default noreply@pescheck.nl)
 */

require __DIR__ . '/vendor/autoload.php';

use Pescheck\Client\Configuration;
use Pescheck\Client\ApiException;
use Pescheck\Client\Api\ChecksApi;
use Pescheck\Client\Api\ProfilesApi;
use Pescheck\Client\Api\ScreeningsApi;
use Pescheck\Client\Api\WebhooksApi;
use Pescheck\Client\Api\DivisionsApi;
use Pescheck\Client\Model\V2ProfileCreate;
use Pescheck\Client\Model\V2ProfileCheck;
use Pescheck\Client\Model\PatchedV2ProfilePartialUpdate;
use Pescheck\Client\Model\V2ScreeningCreate;
use Pescheck\Client\Model\V2Candidate;
use Pescheck\Client\Model\Webhook;
use Pescheck\Client\Model\DivisionWrite;
use Pescheck\Client\Model\PatchedDivisionWrite;

const DIVISION_NAME = 'E2E CI division';

function step(string $msg): void
{
    echo "  - {$msg}\n";
}

$baseUrl = getenv('PESCHECK_BASE_URL') ?: '';
$token = getenv('PESCHECK_ACCESS_TOKEN') ?: '';
$email = getenv('PESCHECK_TEST_EMAIL') ?: 'noreply@pescheck.nl';
if ($baseUrl === '' || $token === '') {
    fwrite(STDERR, "PESCHECK_BASE_URL and PESCHECK_ACCESS_TOKEN must be set\n");
    exit(2);
}

// Profile/webhook names must be unique per run: the API soft-deletes and 500s on
// a duplicate name. The division has no delete, so we reuse a fixed-named one.
$suffix = bin2hex(random_bytes(4));

$config = Configuration::getDefaultConfiguration()->setHost($baseUrl)->setAccessToken($token);
$http = new \GuzzleHttp\Client();
$checks = new ChecksApi($http, $config);
$profiles = new ProfilesApi($http, $config);
$screenings = new ScreeningsApi($http, $config);
$webhooks = new WebhooksApi($http, $config);
$divisions = new DivisionsApi($http, $config);

$profileId = null;
try {
    // --- checks ---
    $checkList = $checks->v2ChecksList();
    $checkType = $checkList[0]->getCheckType();
    $checks->v2ChecksRetrieve($checkType);
    step(sprintf("checks: %d types, retrieved '%s'", count($checkList), $checkType));

    // --- profile: create -> retrieve -> patch ---
    $created = $profiles->v2ProfilesCreate(new V2ProfileCreate([
        'name' => "E2E test profile {$suffix}",
        'description' => 'created by e2e',
        'checks' => [new V2ProfileCheck(['check_type' => $checkType])],
    ]));
    $profileId = $created->getId();
    $profiles->v2ProfilesRetrieve($profileId);
    $profiles->v2ProfilesPartialUpdate($profileId, new PatchedV2ProfilePartialUpdate([
        'description' => 'updated by e2e',
    ]));
    step("profile: created+retrieved+patched ({$profileId})");

    // --- screening: create -> retrieve -> documents ---
    $screening = $screenings->v2ScreeningsCreate(new V2ScreeningCreate([
        'profile_id' => $profileId,
        'candidate' => new V2Candidate([
            'first_name' => 'E2E', 'last_name' => 'Tester', 'email' => $email,
        ]),
    ]));
    $screeningId = $screening->getId();
    $screenings->v2ScreeningsRetrieve($screeningId);
    $screenings->v2ScreeningsDocumentsList($screeningId);
    step("screening: created+retrieved ({$screeningId}, status={$screening->getStatus()})");

    // --- webhook: create -> list -> delete ---
    $hook = $webhooks->createWebhook2(new Webhook([
        'name' => "E2E webhook {$suffix}",
        'url' => "https://example.com/e2e-hook-{$suffix}",
        'events' => ['screening.status_changed'],
    ]));
    $webhooks->listWebhooks2();
    $webhooks->deleteWebhook2($hook->getId());
    step("webhook: created+listed+deleted ({$hook->getId()})");

    // --- division: list -> reuse-or-create -> patch ---
    $existing = $divisions->v2OrganisationsDivisionsList()->getResults();
    $div = null;
    foreach ($existing as $d) {
        if ($d->getName() === DIVISION_NAME) {
            $div = $d;
            break;
        }
    }
    if ($div === null) {
        $div = $divisions->v2OrganisationsDivisionsCreate(new DivisionWrite([
            'name' => DIVISION_NAME, 'city' => 'Amsterdam', 'address' => 'Teststraat 1',
            'postal' => '1011AA', 'phone' => '+31200000000', 'contact_name' => 'E2E',
            'contact_email' => $email, 'invoice_email' => $email,
        ]));
        $action = 'created';
    } else {
        $action = 'reused';
    }
    $divisions->v2OrganisationsDivisionsPartialUpdate($div->getId(), new PatchedDivisionWrite([
        'city' => 'Rotterdam',
    ]));
    step("division: {$action}+patched ({$div->getId()})");

    echo "E2E CRUD OK\n";
    $exit = 0;
} catch (ApiException $e) {
    fwrite(STDERR, "E2E CRUD FAILED: ApiException {$e->getCode()}: {$e->getMessage()}\n");
    fwrite(STDERR, (string) $e->getResponseBody() . "\n");
    $exit = 1;
} catch (\Throwable $e) {
    fwrite(STDERR, "E2E CRUD FAILED: " . get_class($e) . ": {$e->getMessage()}\n");
    $exit = 1;
} finally {
    // best-effort cleanup of the profile we created
    if ($profileId !== null) {
        try {
            $profiles->v2ProfilesDestroy($profileId);
        } catch (\Throwable $e) {
            // ignore
        }
    }
}

exit($exit);

File: examples/java/CrudLifecycle.java
import java.net.URI;
import java.util.List;
import java.util.UUID;

import io.pescheck.client.ApiClient;
import io.pescheck.client.ApiException;
import io.pescheck.client.Configuration;
import io.pescheck.client.auth.OAuth;
import io.pescheck.client.api.ChecksApi;
import io.pescheck.client.api.ProfilesApi;
import io.pescheck.client.api.ScreeningsApi;
import io.pescheck.client.api.WebhooksApi;
import io.pescheck.client.api.DivisionsApi;
import io.pescheck.client.model.V2CheckInfo;
import io.pescheck.client.model.V2ProfileCheck;
import io.pescheck.client.model.V2ProfileCreate;
import io.pescheck.client.model.V2ProfileDetail;
import io.pescheck.client.model.PatchedV2ProfilePartialUpdate;
import io.pescheck.client.model.V2Candidate;
import io.pescheck.client.model.V2ScreeningCreate;
import io.pescheck.client.model.V2ScreeningDetail;
import io.pescheck.client.model.Webhook;
import io.pescheck.client.model.WebhookResponse;
import io.pescheck.client.model.DivisionWrite;
import io.pescheck.client.model.PatchedDivisionWrite;
import io.pescheck.client.model.DivisionReadOnly;

/**
 * End-to-end CRUD lifecycle test for the Pescheck Java client.
 *
 * Exercises the full v2 surface against the API and self-cleans where possible:
 *
 *   checks      list -> retrieve one
 *   profiles    create -> retrieve -> patch -> (appears in list) -> delete
 *   screenings  create (needs a profile + candidate) -> retrieve -> list documents
 *   webhooks    create -> list -> delete
 *   divisions   list -> reuse-or-create by name -> patch
 *
 * Prints a line per step and "E2E CRUD OK" on success; exits non-zero on any error.
 *
 * Configuration via environment variables:
 *   PESCHECK_BASE_URL      base URL of the API (e.g. https://api-staging.pescheck.io)
 *   PESCHECK_ACCESS_TOKEN  OAuth2 client-credentials access token
 *   PESCHECK_TEST_EMAIL    candidate/division email (default noreply@pescheck.nl)
 */
public class CrudLifecycle {

  // Profile/webhook names must be unique per run: the API soft-deletes and returns
  // a 500 on a duplicate name. The division has no delete endpoint, so we reuse a
  // fixed-named one instead of creating a new one each run.
  static final String SUFFIX = UUID.randomUUID().toString().substring(0, 8);
  static final String DIVISION_NAME = "E2E CI division";

  static void step(String msg) {
    System.out.println("  - " + msg);
  }

  public static void main(String[] args) {
    String baseUrl = System.getenv("PESCHECK_BASE_URL");
    String token = System.getenv("PESCHECK_ACCESS_TOKEN");
    String testEmail = System.getenv("PESCHECK_TEST_EMAIL");
    if (testEmail == null || testEmail.isEmpty()) {
      testEmail = "noreply@pescheck.nl";
    }
    if (baseUrl == null || baseUrl.isEmpty() || token == null || token.isEmpty()) {
      System.err.println("PESCHECK_BASE_URL and PESCHECK_ACCESS_TOKEN must be set");
      System.exit(2);
    }

    ApiClient client = Configuration.getDefaultApiClient();
    client.setBasePath(baseUrl);
    // OAuth2 bearer authentication (scheme "oauth2") — the scheme the operations require.
    OAuth oauth2 = (OAuth) client.getAuthentication("oauth2");
    oauth2.setAccessToken(token);

    ChecksApi checks = new ChecksApi(client);
    ProfilesApi profiles = new ProfilesApi(client);
    ScreeningsApi screenings = new ScreeningsApi(client);
    WebhooksApi webhooks = new WebhooksApi(client);
    DivisionsApi divisions = new DivisionsApi(client);

    UUID profileId = null;
    try {
      // --- checks ---
      List<V2CheckInfo> checkList = checks.v2ChecksList(null);
      String checkType = checkList.get(0).getCheckType();
      checks.v2ChecksRetrieve(checkType);
      step("checks: " + checkList.size() + " types, retrieved '" + checkType + "'");

      // --- profile: create -> retrieve -> patch ---
      V2ProfileDetail created = profiles.v2ProfilesCreate(
          new V2ProfileCreate()
              .name("E2E test profile " + SUFFIX)
              .description("created by e2e")
              .checks(List.of(new V2ProfileCheck()
                  .checkType(V2ProfileCheck.CheckTypeEnum.fromValue(checkType)))));
      profileId = created.getId();
      profiles.v2ProfilesRetrieve(profileId);
      profiles.v2ProfilesPartialUpdate(
          profileId, new PatchedV2ProfilePartialUpdate().description("updated by e2e"));
      step("profile: created+retrieved+patched (" + profileId + ")");

      // --- screening: create -> retrieve -> documents ---
      V2ScreeningDetail screening = screenings.v2ScreeningsCreate(
          new V2ScreeningCreate()
              .profileId(profileId)
              .candidate(new V2Candidate()
                  .firstName("E2E").lastName("Tester").email(testEmail)));
      screenings.v2ScreeningsRetrieve(screening.getId());
      screenings.v2ScreeningsDocumentsList(screening.getId(), null, null);
      step("screening: created+retrieved (" + screening.getId()
          + ", status=" + screening.getStatus() + ")");

      // --- webhook: create -> list -> delete ---
      WebhookResponse hook = webhooks.createWebhook2(
          new Webhook()
              .name("E2E webhook " + SUFFIX)
              .url(URI.create("https://example.com/e2e-hook"))
              .events(List.of(Webhook.EventsEnum.fromValue("screening.status_changed"))),
          null, null);
      webhooks.listWebhooks2();
      // webhook id deserializes as a UUID; the delete path param wants a string.
      webhooks.deleteWebhook2(hook.getId().toString());
      step("webhook: created+listed+deleted (" + hook.getId() + ")");

      // --- division: list -> reuse-or-create -> patch ---
      List<DivisionReadOnly> existing =
          divisions.v2OrganisationsDivisionsList(null, null, null).getResults();
      UUID divisionId = null;
      String action;
      for (DivisionReadOnly d : existing) {
        if (DIVISION_NAME.equals(d.getName())) {
          divisionId = d.getId();
          break;
        }
      }
      if (divisionId == null) {
        DivisionWrite div = divisions.v2OrganisationsDivisionsCreate(
            new DivisionWrite()
                .name(DIVISION_NAME).city("Amsterdam").address("Teststraat 1")
                .postal("1011AA").phone("+31200000000").contactName("E2E")
                .contactEmail(testEmail).invoiceEmail(testEmail));
        divisionId = div.getId();
        action = "created";
      } else {
        action = "reused";
      }
      divisions.v2OrganisationsDivisionsPartialUpdate(
          divisionId, new PatchedDivisionWrite().city("Rotterdam"));
      step("division: " + action + "+patched (" + divisionId + ")");

      System.out.println("E2E CRUD OK");
    } catch (ApiException e) {
      System.err.println("E2E CRUD FAILED: ApiException: " + e.getCode() + " " + e.getResponseBody());
      cleanup(profiles, profileId);
      System.exit(1);
    } catch (Exception e) {
      System.err.println("E2E CRUD FAILED: " + e.getClass().getSimpleName() + ": " + e.getMessage());
      cleanup(profiles, profileId);
      System.exit(1);
    }
    // best-effort cleanup of the profile we created
    cleanup(profiles, profileId);
  }

  static void cleanup(ProfilesApi profiles, UUID profileId) {
    if (profileId != null) {
      try {
        profiles.v2ProfilesDestroy(profileId);
      } catch (Exception ignored) {
        // best-effort
      }
    }
  }
}
